### PR TITLE
Make daemon newable

### PR DESCRIPTION
**Pull request by krhubert**

### CODE REVIEW

File: container/service_test.go [hunk 1]
```diff
@@ -13,6 +13,7 @@ import (
 )
 
 func TestStartService(t *testing.T) {
+	t.Skip("put back when dockertest will be replaced with testify.Mock")
```
Comment (ilgooz) [inline container/service_test.go]: why are we skipping tests?

File: container/service_test.go [hunk 2]
```diff
@@ -35,6 +36,8 @@ func TestStartService(t *testing.T) {
 }
 
 func TestStopService(t *testing.T) {
+	t.Skip("put back when dockertest will be replaced with testify.Mock")
```
Comment (ilgooz) [inline container/service_test.go]: why are we skipping tests?

File: container/service.go [hunk 3]
```diff
@@ -39,6 +39,15 @@ func (c *DockerContainer) FindService(namespace []string) (swarm.Service, error)
 
 // StartService starts a docker service.
 func (c *DockerContainer) StartService(options ServiceOptions) (serviceID string, err error) {
+	status, err := c.Status(options.Namespace)
```
Comment (ilgooz) [inline container/service.go]: what is the reason of this change? it is not relevant with the PRs subject.

File: container/service.go [hunk 4]
```diff
@@ -50,7 +59,15 @@ func (c *DockerContainer) StartService(options ServiceOptions) (serviceID string
 }
 
 // StopService stops a docker service.
-func (c *DockerContainer) StopService(namespace []string) (err error) {
+func (c *DockerContainer) StopService(namespace []string) error {
+	status, err := c.Status(namespace)
```
Comment (ilgooz) [inline container/service.go]: what is the reason of this change? it is not relevant with the PRs subject.

File: container/service_test.go (hunk 1 above)
Comment (krhubert) [reply container/service_test.go]: The `dockertest` package doesn't have a Status Method, so I decided to skip tests and put them back when `dockertest` will be replaced with testify mocks (as in description). So I prefer using the time to move to mocks rather then implement the Status method.

File: container/service_test.go (hunk 2 above)
Comment (krhubert) [reply container/service_test.go]: Same as with StartService.

**Comment by ilgooz** [comment]
Do have manual tests done?

**Comment by krhubert** [comment]
yes

File: daemon/daemon_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	c.AssertExpectations(t)
+}
+
+func TestStatus(t *testing.T) {
+	cfg, _ := config.Global()
+	c := &mocks.Container{}
+	d := NewContainerDaemon(cfg, c)
+
+	c.On("Status", []string{}).Return(container.STOPPED, nil)
+	d.Status()
```
Comment (ilgooz) [inline daemon/daemon_test.go]: missing assertion > nil check

File: daemon/daemon_test.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	c.AssertExpectations(t)
+}
+
+func TestStop(t *testing.T) {
+	cfg, _ := config.Global()
+	c := &mocks.Container{}
+	d := NewContainerDaemon(cfg, c)
+
+	c.On("StopService", []string{}).Return(nil)
+	d.Stop()
```
Comment (ilgooz) [inline daemon/daemon_test.go]: missing assertion

File: daemon/daemon_test.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+func TestStart(t *testing.T) {
+	cfg, _ := config.Global()
+	c := &mocks.Container{}
+	d := NewContainerDaemon(cfg, c)
+
+	c.On("SharedNetworkID").Return("1", nil)
+	c.On("StartService", d.buildServiceOptions("1")).Return("1", nil)
+	d.Start()
```
Comment (ilgooz) [inline daemon/daemon_test.go]: missing assertion

File: daemon/daemon_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	c.AssertExpectations(t)
+}
+
+func TestLogs(t *testing.T) {
+	cfg, _ := config.Global()
+	c := &mocks.Container{}
+	d := NewContainerDaemon(cfg, c)
+
+	c.On("ServiceLogs", []string{}).Return(nil, nil)
+	d.Logs()
```
Comment (ilgooz) [inline daemon/daemon_test.go]: missing assertion

File: daemon/daemon_test.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	d.Status()
+	c.AssertExpectations(t)
+}
+
+func TestLogs(t *testing.T) {
+	cfg, _ := config.Global()
+	c := &mocks.Container{}
+	d := NewContainerDaemon(cfg, c)
+
+	c.On("ServiceLogs", []string{}).Return(nil, nil)
```
Comment (ilgooz) [inline daemon/daemon_test.go]: need to return a real log stream to test the behavior

File: daemon/daemon_test.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestStart(t *testing.T) {
+	cfg, _ := config.Global()
+	c := &mocks.Container{}
+	d := NewContainerDaemon(cfg, c)
+
+	c.On("SharedNetworkID").Return("1", nil)
+	c.On("StartService", d.buildServiceOptions("1")).Return("1", nil)
```
Comment (ilgooz) [inline daemon/daemon_test.go]: having _d.buildServiceOptions("1")_ directly is actually does not check possible mistakes in the creation of _container.ServiceOptions_. it should be manually created here.

File: container/service_test.go (hunk 1 above)
Comment (ilgooz) [inline container/service_test.go]: We already have testify mocks, please add test for this one with testify. https://github.com/mesg-foundation/core/pull/588

File: container/service_test.go (hunk 2 above)
Comment (ilgooz) [inline container/service_test.go]: We already have testify mocks, please add test for this one with testify. https://github.com/mesg-foundation/core/pull/588

File: container/service_test.go (hunk 1 above)
Comment (krhubert) [reply container/service_test.go]: It's not merged and I will change all tests at once not one by one.

File: daemon/daemon_test.go (hunk 9 above)
Comment (krhubert) [reply daemon/daemon_test.go]: My comment disapper, but you don't need to return writer, it's just a wrapper and tests are ok here

File: daemon/daemon_test.go (hunk 10 above)
Comment (krhubert) [reply daemon/daemon_test.go]: For buildServiceOptions you have a sperate test. Here it checks if StartService was called
